packages/web/server/api/edge-routers/concentrated-liquidity-router.ts

```diff
@@ -0,0 +1,297 @@
+import { CoinPretty, Dec, DecUtils, Int } from "@keplr-wallet/unit";
+import { tickToSqrtPrice } from "@osmosis-labs/math";
+import cachified, { CacheEntry } from "cachified";
+import { LRUCache } from "lru-cache";
+import { z } from "zod";
+
+import { DEFAULT_LRU_OPTIONS } from "~/config/cache";
+import { createTRPCRouter, publicProcedure } from "~/server/api/trpc";
+import { getAsset } from "~/server/queries/complex/assets";
+import { UserOsmoAddressSchema } from "~/server/queries/complex/parameter-types";
+import { queryPositionPerformance } from "~/server/queries/imperator";
+import { queryAddressCLPositions } from "~/server/queries/osmosis/concentrated";
+
+export const concentratedLiquidityRouter = createTRPCRouter({
+  getUserPositions: publicProcedure
+    .input(
+      z
+        .object({
+          sortDirection: z.enum(["asc", "desc"]).default("desc").optional(),
+        })
+        .merge(UserOsmoAddressSchema.required())
+    )
+    .query(async ({ input: { userOsmoAddress, sortDirection } }) => {
+      const { positions: rawPositions } = await queryAddressCLPositions({
+        bech32Address: userOsmoAddress,
+      });
+
+      // TODO: sort by joinDate
+      const positionsWithPerformance = await getPositionsWithPerformance({
+        userOsmoAddress,
+        positions: rawPositions,
+      });
+      const result = await normalizePositions({
+        positions: positionsWithPerformance,
+      });
+      return result;
+    }),
+});
+
+type LiquidityPosition = Awaited<
+  ReturnType<typeof queryAddressCLPositions>
+>["positions"][number];
+type PositionPerformance = Awaited<ReturnType<typeof queryPositionPerformance>>;
+
+type PositionStatus =
+  | "inRange"
+  | "nearBounds"
+  | "outOfRange"
+  | "fullRange"
+  | "unbonding"
+  | "superfluidStaked"
+  | "superfluidUnstaking";
+
+function getPositionStatus({
+  lowerPrice,
+  upperPrice,
+  currentPrice,
+  isFullRange,
+  isSuperfluid,
+  isSuperfluidUnstaking,
+  isUnbonding,
+}: {
+  lowerPrice: Dec;
+  upperPrice: Dec;
+  currentPrice: Dec;
+  isFullRange: boolean;
+  isSuperfluid: boolean;
+  isSuperfluidUnstaking: boolean;
+  isUnbonding: boolean;
+}): PositionStatus {
+  const inRange = lowerPrice.lt(currentPrice) && upperPrice.gt(currentPrice);
+  const diff = new Dec(
+    Math.min(
+      Number(currentPrice.sub(lowerPrice).toString()),
+      Number(upperPrice.sub(currentPrice).toString())
+    )
+  );
+  const rangeDiff = upperPrice.sub(lowerPrice);
```

Comment:
Could we put these types and fns into complex folder?

Reply:
Should be on their respective folder on my latest commit